Fixes #8007 - Support Loom.

sbordet opened the pull request:
Alternative implementation of support for virtual threads for HTTP/1.1, HTTP/2 and HTTP/3.

The virtual thread support is in AdaptiveExecutionStrategy.
When virtual threads are supported and enabled, reserved threads are disabled and
blocking tasks are run in a virtual thread instead that being executed by the Executor.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #8007 - Support Loom.

Alternative implementation of support for virtual threads for HTTP/1.1, HTTP/2 and HTTP/3.

The virtual thread support is in AdaptiveExecutionStrategy.
When virtual threads are supported and enabled, reserved threads are disabled and
blocking tasks are run in a virtual thread instead that being executed by the Executor.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@gregw @lorban please review this alternative support for Loom, more in line with what suggested by @gregw.

It's not complete as we need to verify that all code paths that call application code are actually invoked with a virtual thread (so more testing needed), but it's functional enough to deserve a review to see if it is the right direction before doing more work.

joakime commented:
Keep in mind that the prior Loom support PR #8035 still exists and is still open

* #8035

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/strategy/AdaptiveExecutionStrategy.java:
```diff
@@ -134,20 +135,27 @@ public class AdaptiveExecutionStrategy extends ContainerLifeCycle implements Exe
      */
     public AdaptiveExecutionStrategy(Producer producer, Executor executor)
+    {
+        this(producer, executor, false);
```
I guess we could have a JVM wide system property (cough splutter gasp!) to configure this default?

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/strategy/AdaptiveExecutionStrategy.java:
```diff
@@ -134,20 +135,27 @@ public class AdaptiveExecutionStrategy extends ContainerLifeCycle implements Exe
         _executor = executor;
-        _tryExecutor = TryExecutor.asTryExecutor(executor);
+        _useVirtualThreads = useVirtualThreads;
+        _tryExecutor = useVirtualThreads ? TryExecutor.NO_TRY : TryExecutor.asTryExecutor(executor);
```
yeah... but no!

If you don't want to use reserved threads, then configure 0 reserved threads.   If reserved threads are configured in the thread pool, then they should be used if available.

No need to do anything special here as the user already can configure a NO_TRY tryExecutor if they want it.

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/strategy/AdaptiveExecutionStrategy.java:
```diff
@@ -462,7 +470,10 @@ private void execute(Runnable task)
         try
         {
-            _executor.execute(task);
+            if (_useVirtualThreads)
+                VirtualThreads.startVirtualThread(task);
+            else
+                _executor.execute(task);
```
Wow, that's pretty simple...
Makes me wonder if we could do it all with just a change to the ThreadPool?   Hmmm but then we'd have virtual threads being selectors.... so nah!

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/VirtualThreads.java:
```diff
@@ -0,0 +1,97 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class VirtualThreads
```
javadoc for this class and its methods.

gregw commented on jetty-quic/quic-server/src/main/java/org/eclipse/jetty/quic/server/QuicServerConnector.java:
```diff
@@ -143,6 +145,25 @@ public void setUseOutputDirectByteBuffers(boolean useOutputDirectByteBuffers)
     }
 
+    @Override
+    public boolean isInvokeApplicationWithVirtualThreads()
```
wrong name

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/Connector.java:
```diff
@@ -97,4 +98,21 @@ public interface Connector extends LifeCycle, Container, Graceful
+
+    public interface VirtualThreadsConfigurable
+    {
+        public default boolean isInvokeApplicationWithVirtualThreads()
```
This is not the right name.
Nor am I convinced if it should be a connection configuration.

gregw reviewed (changes requested):
This is looking much better. 
I've made a couple of comments / niggles.

But I also don't like the way `useVirtualThreads` boolean is passed around.  

Rather than configure connectors,  I'm wondering if it could be a filed/attribute/property associated with the thread pool?  The strategy could then look at the threadpool it has been given and work out how to act from that.

sbordet committed:
Updates after review.

Now virtual thread configuration is done in ThreadPool implementation.
ReservedThreadExecutor heuristics has been changed to take into account virtual thread support, so that by default the number of reserved threads is 0 when using virtual threads.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Added test.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/VirtualThreads.java:
```diff
@@ -0,0 +1,161 @@
+     * supports virtual threads and, if the runtime supports them, whether they are
+     * configured to be used via {@link #setUseVirtualThreads(boolean)}.</p>
+     */
+    public interface Configurable
```
good name!

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/strategy/AdaptiveExecutionStrategy.java:
```diff
@@ -134,19 +135,21 @@ public class AdaptiveExecutionStrategy extends ContainerLifeCycle implements Exe
     {
         _producer = producer;
         _executor = executor;
+        _useVirtualThreads = VirtualThreads.isUseVirtualThreads(executor);
```
Is grabbing this at construction too early?
I know it is nice to have as a final, but I think it is probably more correct to look at the value in `doStart` as it might be set on the executor after the strategy has been built (but not started).

sbordet replied:
Good point, done.

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/strategy/AdaptiveExecutionStrategy.java:
```diff
@@ -462,7 +465,10 @@ private void execute(Runnable task)
         try
         {
-            _executor.execute(task);
+            if (isUseVirtualThreads())
```
Should we make this method protected so that if somebody wants to be more exotic in their logic they can extend?

sbordet replied:
It's already `public`.

gregw reviewed (changes requested):
This looks pretty good.... but other than the niggles I've commented on, you also need to modify `jetty-threadpool.xml` and module to have a property for useVirtualThreads.

Or if you like, you could have a `loom.mod` that depended on thread pool and set the property.

sbordet committed:
Updates after review.

Updated threadpool Jetty module and XML files.

Retrieving whether to use virtual threads from AdaptiveExecutionStrategy.doStart() to allow for late configuration of the thread pool.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/strategy/AdaptiveExecutionStrategy.java:
```diff
@@ -462,7 +471,10 @@ private void execute(Runnable task)
         try
         {
-            _executor.execute(task);
+            if (isUseVirtualThreads())
+                VirtualThreads.startVirtualThread(task);
+            else
+                _executor.execute(task);
```
Should have a test in `AdaptiveExecutionStrategyTest` for this.  It should start a billion threads and be OK :)

sbordet replied:
Billions are a thing of the past. I'll use trillions, because that's what's needed these days with 0.75 CPUs in the cloud.

gregw reviewed (approved):
LGTM other than a specific test for the execution strategy would be good to have.

lorban reviewed (approved):
LGTM
